### PR TITLE
Fixes #4721

### LINKED ISSUE #4721: Dummy atoms next to aromatic are always kekulized even when they should not
### Discussed in https://github.com/rdkit/rdkit/discussions/4716

<div type='discussions-op-text'>

<sup>Originally posted by **laurenreid1** November 19, 2021</sup>
Hi everyone,

I'd like to depict a benzocyclohexane-like compound containing 2 dummy atoms within the cyclohexane ring (e.g c12ccccc1*-*CC2). Using the following code:

```
from rdkit import Chem
from rdkit.Chem import Draw
from rdkit.Chem.Draw import rdMolDraw2D

mol = Chem.MolFromSmiles("c12ccccc1*-*CC2")
dm = Draw.PrepareMolForDrawing(mol)
d2d = rdMolDraw2D.MolDraw2DCairo(240, 135)
d2d.DrawMolecule(dm)
d2d.FinishDrawing()
draw_text = d2d.GetDrawingText()

with open("test.png", "wb") as f:
    f.write(draw_text)

```

I receive this depicition that adds a double bond between the 2 dummies:
image

I believe this is happening during kekulization, where the conjugation of the dummy atoms to the benzene ring is interpreted as aromaticity? I can turn off the behaviour using the following:

```
from rdkit import Chem
from rdkit.Chem import Draw
from rdkit.Chem.Draw import rdMolDraw2D

mol = Chem.MolFromSmiles("c12ccccc1*-*CC2", sanitize=False)
mol.UpdatePropertyCache()
dm = Draw.PrepareMolForDrawing(mol, kekulize=False)
d2d = rdMolDraw2D.MolDraw2DCairo(240, 135)
d2d.drawOptions().prepareMolsBeforeDrawing = False
d2d.DrawMolecule(dm)
d2d.FinishDrawing()
draw_text = d2d.GetDrawingText()

with open("test.png", "wb") as f:
    f.write(draw_text)
```

But this leaves me with the following depiction:

image

Does anyone know if there's a way I can depict the benzene ring in its kekulized form while keeping the bond between the dummy atoms as a single aliphatic bond? I've been playing around with the santizeflags but haven't found a combination that will allow only the benzene bonds to be kekulized.

Thanks,
Lauren

EDIT: My RDKit version is 2020.03.3
</div>

### PR DESCRIPTION
**Pull request by ptosco**
This PR fixes #4721. The fix is straightforward.

### CODE REVIEW

**Comment by greglandrum** [comment]
Hi @ptosco : the logic here fails for molecules where the dummy atom *is* in an aromatic ring but is connected to an atom which is not in one:
```
In [6]: Chem.CanonSmiles('C1CC*2ccccc21')
[04:55:39] Can't kekulize mol.  Unkekulized atoms: 4 5 6 7 8
```

**Comment by ptosco** [comment]
@greglandrum Right, that was _way_ too straightforward to be true :-) I'll rework the PR it as soon as I have a bit of time.

**Comment by greglandrum** [comment]
> @greglandrum Right, that was _way_ too straightforward to be true :-) I'll rework the PR it as soon as I have a bit of time.

yeah, messing with the Kekulization code can't possibly be *that* easy. :-)

**Comment by ptosco** [comment]
@greglandrum Don't be scared off by the number of changes.
Actually the fix is pretty simple and robust, and also computationally efficient (I replaced the `std::find` with a bitset).
All the `RingInfo` changes are about something I have had in mind for a while and now had a good opportunity to implement. I think that having atom/bond ring memberships available can be handy under many circumstances and will avoid users having to looping over `atomRings` and `bondRings` themselves for the same purpose.

**Comment by ptosco** [comment]
@greglandrum I'd like to improve further on this.
I don't think it is correct, as `master` currently does (and this PR does not modify) that
```
Chem.MolFromSmiles("N1****1")
```
resolves to
image
I would like not to see any aromatic bonds.
The same applies to this:
```
Chem.MolFromSmiles("c1ccccc1N1****1")
```
image
I believe that in the absence of at least one non-ambiguously aromatic atom in the ring with dummies, no aromatic bonds should be set, i.e. in both cases I'd like to see a pyrrolidine, not a pyrrole.

@greglandrum Do you agree with me? If so, I'll go ahead.

**Comment by greglandrum** [comment]
@ptosco : yes, I completely agree. There should be one atom which is clearly aromatic

**Comment by ptosco** [comment]
@greglandrum this is now ready for review. In addition to the changes already discussed above I have done a pass of code modernization and cleanup. Apart from stylistic changes (replacement of `boost::tie` with `boost::make_iterator_range`, replacement of a few loops with `stdlib` calls) there are also a couple of optimizations (removed unnecessary `std::map` initializations and two unnecessary copies).
None of the changes breaks any existing test, and the tests that we discussed above and a few more that I thought of while working on this all pass.

File: Code/GraphMol/RingInfo.cpp [hunk 1]
```diff
@@ -16,77 +16,157 @@ RingInfo::INT_VECT RingInfo::atomRingSizes(unsigned int idx) const {
...
-  } else {
-    return 0;
+  if (idx < d_atomMembers.size() && !d_atomMembers[idx].empty()) {
+    return d_atomRings
+        .at(*std::min_element(
+            d_atomMembers[idx].begin(), d_atomMembers[idx].end(),
+            [this](int ri1, int ri2) {
+              return d_atomRings.at(ri1).size() < d_atomRings.at(ri2).size();
+            }))
+        .size();
```
Comment (greglandrum) [inline Code/GraphMol/RingInfo.cpp:48]: this is a "a bit" too much for a one-liner.
how about:
```suggestion
   auto elem = *std::min_element(
            d_atomMembers[idx].begin(), d_atomMembers[idx].end(),
            [this](int ri1, int ri2) {
              return d_atomRings.at(ri1).size() < d_atomRings.at(ri2).size();
            });
    return d_atomRings.at(elem).size();
```
(needs to be clang-formatted)

**Review by greglandrum — CHANGES_REQUESTED** [review]
Some changes suggested.

**Comment by greglandrum** [comment]
@ptosco : sorry it took me so long to get to this one.

File: Code/GraphMol/Aromaticity.cpp [hunk 2]
```diff
@@ -135,7 +132,8 @@ typedef enum {
   TwoElectronDonorType,
   OneOrTwoElectronDonorType,
   AnyElectronDonorType,
-  NoElectronDonorType
+  NoElectronDonorType,
+  LastElectronDonorType = NoElectronDonorType
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/GraphMol/Aromaticity.cpp:136]: Oops, that's a leftover from a previous incarnation of the PR which is not used in the current one. Removed.

File: Code/GraphMol/Aromaticity.cpp [hunk 3]
```diff
@@ -259,50 +249,44 @@ bool incidentNonCyclicMultipleBond(const Atom *at, int &who) {
...
-  const ROMol &mol = at->getOwningMol();
-  ROMol::OEDGE_ITER beg, end;
-  boost::tie(beg, end) = mol.getAtomBonds(at);
-  while (beg != end) {
-    if (!mol.getRingInfo()->numBondRings(mol[*beg]->getIdx())) {
-      if (mol[*beg]->getValenceContrib(at) >= 2.0) {
-        who = mol[*beg]->getOtherAtomIdx(at->getIdx());
+  const auto &mol = at->getOwningMol();
+  for (const auto &bi : boost::make_iterator_range(mol.getAtomBonds(at))) {
+    const auto bond = mol[bi];
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/GraphMol/Aromaticity.cpp:254]: Done.

File: Code/GraphMol/Aromaticity.cpp [hunk 4]
```diff
@@ -259,50 +249,44 @@ bool incidentNonCyclicMultipleBond(const Atom *at, int &who) {
...
   PRECONDITION(at, "bad atom");
-  const ROMol &mol = at->getOwningMol();
-  ROMol::OEDGE_ITER beg, end;
-  boost::tie(beg, end) = mol.getAtomBonds(at);
-  while (beg != end) {
-    if (mol.getRingInfo()->numBondRings(mol[*beg]->getIdx())) {
-      if (mol[*beg]->getValenceContrib(at) >= 2.0) {
+  const auto &mol = at->getOwningMol();
+  for (const auto &bi : boost::make_iterator_range(mol.getAtomBonds(at))) {
+    const auto bond = mol[bi];
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/GraphMol/Aromaticity.cpp:269]: Done.

File: Code/GraphMol/Aromaticity.cpp [hunk 5]
```diff
@@ -259,50 +249,44 @@ bool incidentNonCyclicMultipleBond(const Atom *at, int &who) {
...
 bool incidentMultipleBond(const Atom *at) {
   PRECONDITION(at, "bad atom");
+  const auto &mol = at->getOwningMol();
   int deg = at->getDegree() + at->getNumExplicitHs();
-  ROMol::OEDGE_ITER beg, end;
-  boost::tie(beg, end) = at->getOwningMol().getAtomBonds(at);
-  while (beg != end) {
-    Bond *bond = at->getOwningMol()[*beg];
+  for (const auto &bi : boost::make_iterator_range(mol.getAtomBonds(at))) {
+    const auto bond = mol[bi];
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/GraphMol/Aromaticity.cpp:284]: Done.

File: Code/GraphMol/RingInfo.cpp [hunk 6]
```diff
@@ -16,77 +16,157 @@ RingInfo::INT_VECT RingInfo::atomRingSizes(unsigned int idx) const {
   PRECONDITION(df_init, "RingInfo not initialized");
 
   if (idx < d_atomMembers.size()) {
-    return d_atomMembers[idx];
-  } else {
-    return INT_VECT{0};
+    INT_VECT res(d_atomMembers[idx].size());
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/GraphMol/RingInfo.cpp:19]: Good point; I have also exposed them to Python.

File: Code/GraphMol/Kekulize.cpp [hunk 7]
```diff
@@ -100,10 +113,14 @@ void markDbondCands(RWMol &mol, const INT_VECT &allAtms,
     // bonds that we will later mark as being single:
     int sbo = 0;
     unsigned nToIgnore = 0;
-    RWMol::OEDGE_ITER beg, end;
-    boost::tie(beg, end) = mol.getAtomBonds(at);
-    while (beg != end) {
-      Bond *bond = mol[*beg];
+    unsigned int nonArNonDummyNbr = 0;
+    for (const auto &bi : boost::make_iterator_range(mol.getAtomBonds(at))) {
+      auto bond = mol[bi];
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/GraphMol/Kekulize.cpp:118]: Done

File: Code/GraphMol/Kekulize.cpp [hunk 8]
```diff
@@ -132,10 +154,11 @@ void markDbondCands(RWMol &mol, const INT_VECT &allAtms,
...
 
       sbo += at->getTotalNumHs();
-      int dv = PeriodicTable::getTable()->getDefaultValence(at->getAtomicNum());
-      int chrg = at->getFormalCharge();
+      auto dv =
+          PeriodicTable::getTable()->getDefaultValence(at->getAtomicNum());
+      auto chrg = at->getFormalCharge();
       if (isEarlyAtom(at->getAtomicNum())) {
-        chrg *= -1;  // fix for GitHub #65
+        chrg = -chrg;  // fix for GitHub #65
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/GraphMol/Kekulize.cpp:161]: We have already had the argument on performance once and I lost it (I happily lose those arguments as I actually hate to argue, particularly on unimportant things: all the pleasure is in disagreeing), so I won't claim it's for performance reasons, it's purely for consistency with what is used two lines below:
https://github.com/rdkit/rdkit/blob/b19af71acd7d87a5f72301f0808baf72b5fd405e/Code/GraphMol/Kekulize.cpp#L142
I can't lose this one, can I ;-)?

[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/Kekulize.cpp:161]: hmmm... looks like you got me there. :-)

File: Code/GraphMol/RingInfo.cpp (hunk 1 above)
Comment (ptosco) [reply Code/GraphMol/RingInfo.cpp:48]: Done.

File: Code/GraphMol/RingInfo.cpp [hunk 9]
```diff
@@ -16,77 +16,157 @@ RingInfo::INT_VECT RingInfo::atomRingSizes(unsigned int idx) const {
...
+
+  static const INT_VECT emptyVect;
+  if (idx < d_atomMembers.size()) {
+    return d_atomMembers[idx];
+  }
+  return emptyVect;
+}
+bool RingInfo::areAtomsInSameRingInternal(unsigned int idx1, unsigned int idx2,
+                                          bool useSize,
+                                          unsigned int size) const {
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/GraphMol/RingInfo.cpp:71]: That makes sense, done.

**Comment by ptosco** [comment]
@greglandrum I should have addressed all your remarks; please shou if that’s not the case. Thanks for reviewing!

**Comment by greglandrum** [comment]
@ptosco since making behavioral changes to core pieces like kekulization makes me nervous, I tried a few things.

This behavior doesn't feel right to me and I'm wondering if you can explain why it is right:
image

**Comment by ptosco** [comment]
@greglandrum It makes me nervous too, be assured, and I am totally for being conservative for low-level algorithms like kekulization/aromaticity as it is easy to break things. However, we should consider that the way the current aromaticity code deals with dummy atoms is not satisfactory (see examples above where structures are arbitrarily aromatized) and we should improve it if possible, rather than simply sticking to an unsatisfactory status quo.
That said, I can try and explain my reasoning.

In the first example:
```
m = Chem.MolFromSmiles('C1=C*2=CC=CC=*2C=C1')
m
```
image
```
for b in m.GetBonds():
    print (b.GetBeginAtomIdx(), b.GetEndAtomIdx(), b.GetBondType())
0 1 AROMATIC
1 2 AROMATIC
2 3 AROMATIC
3 4 AROMATIC
4 5 AROMATIC
5 6 AROMATIC
6 7 AROMATIC
7 8 AROMATIC
8 9 AROMATIC
9 0 AROMATIC
7 2 AROMATIC
```
All 5 double bonds were already positioned in the SMILES, the system is fully conjugated and satisfies Hückel, so it is labelled as aromatic.

In the second example
```
m = Chem.MolFromSmiles('C1=C*2C=CC=C*2C=C1')
m
```
image

```
for b in m.GetBonds():
    print (b.GetBeginAtomIdx(), b.GetEndAtomIdx(), b.GetBondType())
0 1 DOUBLE
1 2 SINGLE
2 3 SINGLE
3 4 DOUBLE
4 5 SINGLE
5 6 DOUBLE
6 7 SINGLE
7 8 SINGLE
8 9 DOUBLE
9 0 SINGLE
7 2 SINGLE
```
only 4 double bonds were positioned. With the available information, it is not possible to guess if the user wished the dummy atoms to be sp3 or sp2-hybridized. As there is no certainty on the intention of the user to represent an aromatic system or rather a bicyclotetraene, the system is left as is, as we discussed above.
If the user explicitly marks the system as aromatic, fully kekulizing it or explicitly marking atoms as aromatic, things will work as expected:

```
m = Chem.MolFromSmiles('c1c*2cccc*2cc1')
m
```
image

```
for b in m.GetBonds():
    print (b.GetBeginAtomIdx(), b.GetEndAtomIdx(), b.GetBondType())
0 1 AROMATIC
1 2 AROMATIC
2 3 AROMATIC
3 4 AROMATIC
4 5 AROMATIC
5 6 AROMATIC
6 7 AROMATIC
7 8 AROMATIC
8 9 AROMATIC
9 0 AROMATIC
7 2 AROMATIC
```
The idea is that kekulization is inhibited unless there is certainty on the aromatic nature of the dummies. That seems a reasonable assumption to me, but I may certainly be wrong, and I am looking forward to hearing your opinion.

**Comment by greglandrum** [comment]
@ptosco : that explanation makes perfect sense to me and I'm :+1: on the logic

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by bp-kelley** [comment]
Sorry for the lateness, but we did have some code to do this in the reaction sanitizer, in aromatize if possible for smarts patterns.

```
>>> rxn = AllChem.ReactionFromSmarts("C1=C*2C=CC=C*2C=C1>>C")
>>> Chem.MolToSmiles(rxn.GetReactantTemplate(0))
'C1=C*2C=CC=C*2C=C1'
>>> AllChem.SanitizeRxn(rxn)
[11:05:53] Mismatched potential rlabels: 2 unmapped reactant dummy atom rlabels,0 unmappped product dummy atom rlabels
rdkit.Chem.rdChemReactions.SanitizeFlags.SANITIZE_NONE
>>> Chem.MolToSmiles(rxn.GetReactantTemplate(0))
'c1cc*2cccc*:2c1'
```

I'm just curious about the differences here.

**Comment by ptosco** [comment]
@bp-kelley Hi Brian, if `params.aromatizeIfPossible` is `true` (as it is, by default), `adjustQueryProperties()` calls `sanitizeMol` to set aromatic flags:
```
  if (params.aromatizeIfPossible) {
    unsigned int failed;
    sanitizeMol(mol, failed, SANITIZE_SYMMRINGS | SANITIZE_SETAROMATICITY);
  } else {
[...]
```
After this PR has been merged, sanitizing with `SANITIZE_SETAROMATICITY` does not aromatize the structure anymore, as it is not obvious if the user meant to treat the dummies as sp2 (hence aromatic) or sp3 (non-aromatic):
```
rxn = rdChemReactions.ReactionFromSmarts("C1=C*2C=CC=C*2C=C1>>C")
rxn
```
image

```
Chem.MolToSmiles(rxn.GetReactantTemplate(0))
'C1=C*2C=CC=C*2C=C1'
rdChemReactions.SanitizeRxn(rxn)
[17:25:08] Mismatched potential rlabels: 2 unmapped reactant dummy atom rlabels,0 unmappped product dummy atom rlabels
rdkit.Chem.rdChemReactions.SanitizeFlags.SANITIZE_NONE
Chem.MolToSmiles(rxn.GetReactantTemplate(0))
'C1=C*2C=CC=C*2C=C1'
```
However, if you add an explicit double bond between the dummies (or use aromatic atoms/bonds in the SMARTS), the ambiguity is solved and therefore the structure will be aromatized upon sanitization:
```
rxn = rdChemReactions.ReactionFromSmarts("C1=C*2C=CC=C*=2C=C1>>C")
rxn
```
image

```
Chem.MolToSmiles(rxn.GetReactantTemplate(0))
'C1=C*2=*(C=C1)C=CC=C2'
rdChemReactions.SanitizeRxn(rxn)
[17:34:19] Mismatched potential rlabels: 2 unmapped reactant dummy atom rlabels,0 unmappped product dummy atom rlabels
rdkit.Chem.rdChemReactions.SanitizeFlags.SANITIZE_NONE
Chem.MolToSmiles(rxn.GetReactantTemplate(0))
'c1cc*2cccc*:2c1'
```
So the behavior is consistent with what described in my [comment above](https://github.com/rdkit/rdkit/pull/4722#issuecomment-1024297176).

**Comment by greglandrum** [comment]
@bp-kelley I think it’s important to remember (and I had to keep reminding myself of this) that the PR only changes the behavior for rings which contain dummy atoms. And then only for some of those, as @ptosco describes above.

**Pull request merged**